### ISSUE TITLE
Potential Bug in the Ipopt-Watertap Solver factory 

### ISSUE THREAD

**Issue by aladshaw3** [watertap-org/watertap#387]
I came across this strange error (on windows) when trying to test out the development of a new unit model. I had done this same test on Linux (about a week ago) and it ran everything fine, but now on Windows the code gets caught in an infinite loop of errors whenever I attempt to initialize or solve the model. 

I was able to narrow it down to the solver factor for 'ipopt-watertap'. If I just use the standard solver factory from IDAES (i.e., change 'ipopt-watertap' to 'ipopt', the code runs fine. 

Here is how to recreate the issue:

(1) My basic unit model (this model is incomplete, but enough is there to run a solve on): https://github.com/aladshaw3/watertap/blob/solver-issue/watertap/unit_models/electrodialysis_0D.py 

(2) File to run the unit model: https://github.com/aladshaw3/watertap/blob/solver-issue/watertap/unit_models/run_elec_example.py 

In that above file (2) there is a boolean I use to change the solver from 'ipopt-watertap' to 'ipopt'. If set to false, it uses 'ipopt-watertap'. Setting that to true uses 'ipopt'. 

If you run that file as is (with the boolean set to false), then you get the following errors.
[ipopt-watertap-error.txt](https://github.com/watertap-org/watertap/files/8189430/ipopt-watertap-error.txt)

However, if you set that boolean to true (i.e., just use the default solver of IDAES, 'ipopt'), then the model initializes and solves fine.

**Comment by bknueven** [comment]
This example runs fine on my system (Linux).

It looks to be an issue with complied libraries in an upstream package. What is the output of:
- `pip list`
- `python -c "from pyomo.contrib.pynumero.asl import AmplInterface; print(AmplInterface.available())"` ?

**Comment by aladshaw3** [comment]
The example also runs fine on my linux machine. For whatever reason, this issue seems isolated to windows?

Here are the outputs of pip-list for windows and linux.

[pip-list-windows.txt](https://github.com/watertap-org/watertap/files/8199431/pip-list-windows.txt)
[pip-list-linux.txt](https://github.com/watertap-org/watertap/files/8199432/pip-list-linux.txt)

For the second command you listed, both windows and linux actually return 'False' for me.

**Comment by aladshaw3** [comment]
Also, I have already re-installed watertap on both machines and have created a clean conda environment, but still have the same issues with Windows.

**Comment by bknueven** [comment]
The second command should print `True`. Have you run `idaes get-extensions` recently on either machine?

**Comment by lbianchi-lbl** [comment]
@aladshaw3 you can also try `idaes environment-info`, it should display some information about the solvers/`idaes-ext` versions that are available on the system.

**Comment by bknueven** [comment]
@lbianchi-lbl I'm a bit surprised that doesn't print the version of the idaes extension package? Is there a recommended way to get that? Obviously `cat ~/.idaes/bin/version_lib.txt` works for my system, but not on Windows.

I suspect @aladshaw3's PyNumero extensions is out-of-sync with his Pyomo version, and for some reason this is okay on Linux but not Windows. `ipopt-watertap` calls PyNumero *though* IDAES, and based on the stack trace this is where things are going wrong.

**Comment by lbianchi-lbl** [comment]
@bknueven maybe `idaes get-extensions --show-current-version`?

**Comment by bknueven** [comment]
> @bknueven maybe `idaes get-extensions --show-current-version`?

@aladshaw3 I would be curious what your system currently shows for this command.

**Comment by aladshaw3** [comment]
> > @bknueven maybe `idaes get-extensions --show-current-version`?
> 
> @aladshaw3 I would be curious what your system currently shows for this command.


(watertap-dev) PS C:\Users\1pi\projects\watertap> idaes get-extensions --show-current-version
---------------------------------------------------
IDAES Extensions Build Versions
===================================================
Solvers:  v2.5.5 20220115-1054 windows
Library:  v2.5.5 20220115-1056 windows
===================================================

**Comment by aladshaw3** [comment]
Here is the results of 'idaes environment-info'
[idaes-env-info.txt](https://github.com/watertap-org/watertap/files/8199807/idaes-env-info.txt)

**Comment by bknueven** [comment]
@aladshaw3 do you still get `False` when running `python -c "from pyomo.contrib.pynumero.asl import AmplInterface; print(AmplInterface.available())"`?

**Comment by aladshaw3** [comment]
> @aladshaw3 do you still get `False` when running `python -c "from pyomo.contrib.pynumero.asl import AmplInterface; print(AmplInterface.available())"`?

Yes, it still returns false for me.

**Comment by aladshaw3** [comment]
I should also note that all the tests pass as well in the environment.

**Comment by lbianchi-lbl** [comment]
@bknueven I'm reading a bit about similar errors and from what I understand it's an intrinsic limitation of using `multiprocessing` on Windows: code using `multiprocessing` must be run as `def main(): ... if __name__ == "__main__": main()`.

In other words (again, from what I understand), it's a "pick any two" situation where the options are:

- Windows
- The ability to run Python code at the module level, as a script (without the `if __name__ ...`)
- Using `multiprocessing`

If this is the case (once more, this is mostly a guess, and I could very well be wrong), since Windows support is not going anywhere, we could see if it's possible to configure or alter the Pyomo code that loads the DLLs so that it doesn't use `multiprocessing`. Maybe the `multiprocessing` implementation was added at some point as an optimization at some point (complete speculation on my side) and it can be turned off or reverted in some way if it's not essential.

**Comment by lbianchi-lbl** [comment]
Also (and I swear this is not a personal grudge against running code at the import level), the error message phrases the `if __name__ ...` syntax as "the *proper* idiom" (emphasis mine):

```txt
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

Therefore, I'm wondering if trying to keep the ability of running "`__main__`-less" wouldn't be going "against the grain" (at least as far as Windows is concerned), and possibly set us up for other similar gotchas in the future.

**Comment by lbianchi-lbl** [comment]
Looking at the Pyomo code, it seems that the reason why `multiprocessing` is used is to enforce a timeout to account for cases where the DLL import might hang:

https://github.com/Pyomo/pyomo/blob/fb0ceec8f662d9e19eef07b30d19be3b25a406e7/pyomo/common/env.py#L42-L81

If `multiprocessing` is being used only for the timeout (as opposed to something for which threads can't be used, e.g. GIL limitations for CPU-bound code), I wonder if using `threading` instead of `multiprocessing` would be feasible, as that would not cause the `RuntimeError` when called from a `__main__`-less module context.

**Comment by bknueven** [comment]
@lbianchi-lbl thanks for doing a deep-dive on this. I now recall that I first came across this particular issue probably 5+ years ago.

I did a cursory search -- outside of testing this seems to be the only code in Pyomo which uses `multiprocessing`. I wonder if its worth trying to come up with patch for Pyomo for just this specific functionality. It would seem like we can use the `join` method on `Thread` to achieve the same behavior: https://docs.python.org/3.8/library/threading.html#threading.Thread.join. However, it's possible that the fact that the interpreter "hangs" when executing this code causes the launched thread to not release the GIL. I don't really have any experience with threading in Python because its rarely useful in a scientific computing context (thanks to the GIL).

At this point I'm also happy just to document this caveat for Windows. If anything, I was planning on making `ipopt-watertap` *more*, not *less*, dependent on PyNumero in the future (see #298). Because WaterTAP often deals with tiny quantities, the alternatives are less appealing.
